### DISCUSSION TITLE
ESP32 ADC & DAC Example

### DISCUSSION THREAD

**Discussion by andrethomas** [pschatzmann/arduino-audio-tools#541] [Q&A] [answered]
Is there an example to use two analog inputs on the ESP32 say ADC1_6 (GPIO32) and ADC2_5 (GPIO14) for left and right audio and then output the audio using the DAC on GPIO25 and GPIO26 respectively?

I looked at adc-serial.ino but I am not sure how to pick up the second ADC and how to send the audio out again using the built in DAC.

Thanks

**Comment by pschatzmann** [comment, marked as answer]
I don't see this as an issue but rather a discussion because my library works as designed!
There is not example because the ESP32 does not support this scenario. I2S using the ADC and the DAC work only on port 0. So you can't  use them at the same time. 
If you use an external DAC and ADC there are not such limitations...

**Comment by andrethomas** [comment]
Thanks, and now I learnt something also...

**Comment by andrethomas** [comment]
How about using something that has two I2S like a stm32f407vet6?

**Comment by pschatzmann** [comment]
The ESP32 has also 2 I2S ports and you can use both of them full duplex and read and write at the same time!
You could also try to use the ADC on port 0 and use the I2S PDM output on port 1.

So far I was only working with a Black Pill STM32411, so feel free to extend/adapt the functionality...

**Comment by leca457** [comment]
Maybe I'm a bit late, but what about just read audio from, say, analog Mic (for example, MAX4466) on an ADC pin and then just copy that audio to DAC pin 34, for example? I've tried to do that, but it is not working. Should I open a new issue and discuss it?

**Reply by pschatzmann** [reply]
The ESP32 does not support this scenario for the same reason mentioned above! 
It would work if you replace the DAC output with PWM or I2S.

**Reply by leca457** [reply]
Hm. Is there a difference between PWM and DAC in audio quality? Thanks for a quick answer tho!

**Reply by pschatzmann** [reply]
Yes, but the signal from the ADC is usually quite bad, so it might be good enough.

**Reply by leca457** [reply]
I get it. I've just set up the things with ADC input and PWM out and I hear nearly no sounds from my speaker. I can hear some very quiet hiss that becomes a bit louder when I blow in my mic. 
In general, I'd like to make a walkie-talkie with esp32 and nRF24L1. Is there anything that you could advise? I should buy I2S mic & speaker or for this project PWM and ADC could be enough? Thanks a lot!!

**Reply by pschatzmann** [reply]
What sample rate did you use ?  I think 8000 should be good enough for this purpose.

The easiest would be to replace the analog microphone with an I2S mirophone: This should increase the quality considerably and you can still use the internal DAC.

As an alternative you can try to use an AnalogAudioStream with the AnalogDriverArduino which is using a timer together with the standard Arduino analog input (or output). This should work since it is not using the DMA...

**Reply by leca457** [reply]
Aha. Thanks. I'll look into that!

**Reply by pschatzmann** [reply]
I restructured the code a bit to make the AnalogAudioArduino class available for the ESP32. I did not test this however...
